linkerd/proxy/http/src/client.rs

```diff
@@ -160,46 +168,51 @@ impl<C: Clone, B> Clone for MakeClient<C, B> {
 
 impl<C, T, B> Future for MakeFuture<C, T, B>
 where
-    C: tower::MakeConnection<T> + Send + Sync + 'static,
-    C::Connection: Send + 'static,
+    C: tower_03::make::MakeConnection<T> + Unpin + Send + Sync + 'static,
+    C::Connection: Unpin + Send + 'static,
     C::Future: Send + 'static,
     C::Error: Into<Error>,
-    B: hyper::body::Payload + 'static,
+    B: hyper::body::HttpBody + Send + 'static,
+    B::Data: Send,
+    B::Error: std::error::Error + Send + Sync,
 {
-    type Item = Client<C, T, B>;
-    type Error = Error;
+    type Output = Result<Client<C, T, B>, Error>;
 
-    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
-        let svc = match *self {
-            MakeFuture::Http1(ref mut h1) => Client::Http1(h1.take().expect("poll more than once")),
-            MakeFuture::Http2(ref mut h2) => {
-                let svc = try_ready!(h2.poll());
+    #[project]
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        #[project]
```

Comment:
For my understanding, what does this inner `#[project]` accomplish?

Reply:
The `#[project]` attribute [allows matching on projected enum variants](https://docs.rs/pin-project/0.4.16/pin_project/attr.project.html#match-expressions). Since attribute macros on expressions (such as `match`) are not available on stable, we have to also place it on the function so that `pin-project` can pre-process the function body to find the attribute.